sdk/opendp/whitenoise/data/dpsu.py

```diff
@@ -0,0 +1,117 @@
+import pandas as pd
+import math
+import random
+import operator
+from collections import defaultdict
+
+from opendp.whitenoise.metadata import CollectionMetadata
+from opendp.whitenoise.sql import PandasReader
+from opendp.whitenoise.sql.parse import QueryParser
+from opendp.whitenoise.mechanisms.rand import laplace
+
+def preprocess_df_from_query(meta_path, csv_path, query, table_name):
+	"""
+	Returns a dataframe with user_id | tuple based on query grouping keys.
+	"""
+	schema = CollectionMetadata.from_file(meta_path)
+	df = pd.read_csv(csv_path)
+
+	reader = PandasReader(schema, df)
+	queries = QueryParser(schema).queries(query)
+	query_ast = queries[0]
+	
+	group_cols = [ge.expression.name for ge in query_ast.agg.groupingExpressions]
+	kc = schema[table_name].key_cols()[0].name
+
+	preprocessed_df = pd.DataFrame()
+	preprocessed_df[kc] = df[kc]
+	preprocessed_df["group_cols"] = tuple(df[group_cols].values.tolist())
+
+	return preprocessed_df
+
+def reservoir_sample(iterable, max_contrib):
+	reservoir = []
+	for i, item in enumerate(iterable):
+		if i < max_contrib:
```

Comment:
python usually tabs in by 4 spaces not 8

Reply:
+1